File: modules/ggpoint.R
#' Shiny Server for Geyser Ggplot2 Point Plot
#'
#' @param id shiny identifier
#' @param df data frame
#' @param xvar,yvar x and y columns of data frame `df`

#' @return reactive server
#' @export
#' @rdname ggpointServer
#' @importFrom shiny bootstrapPage checkboxInput moduleServer NS plotOutput
#'             renderPlot renderUI selectInput shinyApp sliderInput uiOutput
#' @importFrom ggplot2 aes geom_point geom_rug geom_smooth ggplot ggitlel
#'             xlab ylab
#' @importFrom rlang .data
ggpointServer <- function(id, df, xvar, yvar) {
  shiny::moduleServer(id, function(input, output, session) {
    ns <- session$ns
    
    # Output Main Plot
    output$main_plot <- shiny::renderPlot({
      p <- ggplot2::ggplot(df) +
        ggplot2::aes(.data[[xvar]], .data[[yvar]]) +
        ggplot2::geom_point(
          color = "black", fill = "white") +
        ggplot2::xlab("Duration (minutes)") +
        ggplot2::ylab("Wait time (minutes)") +
        ggplot2::ggtitle("Geyser eruption duration and wait time")
      
      if (input$individual_obs) {
        p <- p + ggplot2::geom_rug()
      }
      if (input$smooth) {
        shiny::req(input$bw_adjust)
        p <- p +
          ggplot2::geom_smooth(formula = "y~x", se = FALSE,
            method = "loess", span = input$bw_adjust,
            color = "blue", linewidth = 1)
      }
      print(p)
    })
    
    # Input Bandwidth Adjustment
    output$bw_adjust <- shiny::renderUI({
      if(input$smooth) {
        shiny::sliderInput(inputId = ns("bw_adjust"),
                           label = "Span adjustment:",
                           min = 0, max = 2, value = 0.75, step = 0.05)
      }
    })
  })
}
#' Shiny Module Input for Geyser Ggplot2 Histogram
#' @param id identifier for shiny reactive
#' @return nothing returned
#' @rdname ggpointServer
#' @export
ggpointInput <- function(id) {
  ns <- shiny::NS(id)
  shiny::tagList(
    shiny::checkboxInput(inputId = ns("individual_obs"),
                  label = shiny::strong("Show individual observations"),
                  value = FALSE),
    
    shiny::checkboxInput(inputId = ns("smooth"),
                  label = shiny::strong("Show smooth estimate"),
                  value = FALSE))
}
#' Shiny Module UI for Geyser Ggplot2 Histogram
#' @param id identifier for shiny reactive
#' @return nothing returned
#' @rdname ggpointServer
#' @export
ggpointUI <- function(id) {
  ns <- shiny::NS(id)
  shiny::uiOutput(ns("bw_adjust"))
}
#' Shiny Module Output for Geyser Ggplot2 Histogram
#' @param id identifier for shiny reactive
#' @return nothing returned
#' @rdname ggpointServer
#' @export
ggpointOutput <- function(id) {
  ns <- shiny::NS(id)
  shiny::plotOutput(ns("main_plot"), height = "300px")
}
#' Shiny Module App for Geyser Ggplot2 Histogram
#' @return nothing returned
#' @rdname ggpointServer
#' @export
ggpointApp <- function() {
  ui <- shiny::bootstrapPage(
    ggpointInput("ggpoint"), 
    ggpointOutput("ggpoint"),
    # Display this only if the smooth is shown
    ggpointUI("ggpoint")
  )
  server <- function(input, output, session) {
    ggpointServer("ggpoint", faithful, "eruptions", "waiting")
  }
  shiny::shinyApp(ui, server)
}

File: geyser/ggpoint.py
from shiny import App, module, reactive, render, ui
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from plotnine import ggplot, aes, after_stat, geom_point, geom_rug
from plotnine import geom_smooth, xlab, ylab, ggtitle
from scipy.stats import gaussian_kde
import geyser.io as io

# Create default reactive for dataset.
@reactive.calc
def default_dataset():
    return io.r_object('faithful')

@module.server
def ggpoint_server(input, output, session, data_set=default_dataset):
    """GGPoint Server."""

    if data_set is None:
        return None
    
    @reactive.calc
    def xval():
        return data_set().columns[0]
    def yval():
        return data_set().columns[1]

    @render.plot
    def main_plot():
        if not isinstance(data_set(), pd.DataFrame):
            return None
        if data_set().shape[1] < 2:
            return None
        p = (ggplot(data_set()) +
            aes(x = xval(), y = yval()) +
            geom_point(color = "black", fill = "white"))

        if input.individual_obs():
            p = p + geom_rug()

        if input.density():
            bw_adjust = input.bw_adjust()
            p = p + geom_smooth(color = "blue", span = bw_adjust)

        p = (p +
            xlab(xval()) + ylab(yval()) +
            ggtitle(xval()))
        
        return p
        
    @render.ui
    def output_bw_adjust():
        if input.density():
            return ui.input_slider(
                "bw_adjust",
                "Bandwidth adjustment:",
                min=0,
                max=1.0,
                value=0.5,
                step=0.05
            )
    
# ggpoint_server("ggpoint")

@module.ui
def ggpoint_input():
    """ggpoint Input."""
    return ui.card(
        ui.input_checkbox(
            "individual_obs",
            "Show individual observations",
            value=False
        ),
        ui.input_checkbox(
            "density",
            "Show density estimate",
            value=False
        )
    )

# ggpoint_input("ggpoint")

@module.ui
def ggpoint_output():
    """ggpoint Output."""
    return ui.output_plot("main_plot")

# ggpoint_output("ggpoint")


@module.ui
def ggpoint_ui():
    """ggpoint UI."""
    return ui.output_ui("output_bw_adjust")

# ggpoint_ui("ggpoint")

def ggpoint_app():
    """ggpoint App."""
    app_ui = ui.page_fluid(
        ggpoint_input("ggpoint"),
        ggpoint_output("ggpoint"),
        ggpoint_ui("ggpoint")
    )
    def app_server(input, output, session):
        ggpoint_server("ggpoint")

    app = App(app_ui, app_server)

    #if __name__ == '__main__':
    io.app_run(app)

# ggpoint_app()
